src/Pos/Genesis.hs

```diff
@@ -109,35 +109,73 @@ stakeDistribution ts@RichPoorStakes {..} =
         else identity
     basicDist = genericReplicate sdRichmen sdRichStake ++
                 genericReplicate sdPoor sdPoorStake
-stakeDistribution (CustomStakes coins) = map (,[]) coins
-
--- | Generates genesis 'Utxo' given optional boot stakeholders list
--- and address distribution. In case boot stakeholders are supplied,
--- all the balance is distributed among them. Otherwise, txDistr is
--- set to @[]@ or left as it is (in case of 'ExplicitStakes').
-genesisUtxo :: Maybe (HashMap StakeholderId Word16) -> [AddrDistribution] -> Utxo
-genesisUtxo bootStakeholders ad =
-    M.fromList $ concatMap (uncurry toUtxo . second stakeDistribution) ad
+stakeDistribution (CustomStakes coins) = coins
+
+-- | Generates genesis 'Utxo' given optional boot stakeholders (with
+-- weights) and address distributions.  If genesis stakeholders are
+-- not supplied, they are calculated from address distributions by
+-- making each 'PubKeyAddress' genesis stakeholder with weight equal
+-- to balance. All the stake is distributed among genesis stakeholders
+-- (using 'genesisSplitBoot').
+--
+-- TODO [CSL-1351] This documentation will become correct later.
+genesisUtxo ::
+       Maybe (HashMap StakeholderId Word16) -> [AddrDistribution] -> GenesisUtxo
+genesisUtxo bootStakeholdersMaybe ad
+    -- TODO [CSL-1351] Uncomment ↓.
+    --- | null bootStakeholders =
+    ---     error "genesisUtxo: no stakeholders for the bootstrap era"
+    -- TODO [CSL-1351] Delete (it's here to please hlint) ↓.
+    | ((*) @Int) 2 3 == 7 = error "hlint — makaka"
+    | otherwise = GenesisUtxo . M.fromList $ map utxoEntry balances
   where
-    defaultStakeDistr coin distr
-        = maybe distr
-                (\genDistr -> genesisSplitBoot genDistr coin)
-                bootStakeholders
-    toUtxo :: [Address] -> [(Coin, TxOutDistribution)] -> [(TxIn, TxOutAux)]
-    toUtxo addr distrs =
-        let utxoEntry a (coin,txOutDistr) =
-                ( TxIn (unsafeHash a) 0
-                , TxOutAux (TxOut a coin) (defaultStakeDistr coin txOutDistr)
-                )
-        in zipWith utxoEntry addr distrs
+    -- This type is drunk.
+    somethingComplicated :: [([Address], [Coin])]
+    somethingComplicated = map (second stakeDistribution) ad
+    balances :: [(Address, Coin)]
+    balances = concatMap (uncurry zip) somethingComplicated
+    utxoEntry (addr, coin) =
+        ( TxIn (unsafeHash addr) 0
+-- Note: it's quite bad because HD wallets become stakeholders and each
+-- transaction will give them quite a lot of stake.
```

Comment:
I don't want to rephrase it in the code, because I expect you to do CSL-1351 soon. I can clarify here.
Genesis addresses in dev mode consist of normal `PubKey` addresses and HD addresses (see `devAddrDistr`). HD addresses shouldn't have much stake and currently there is this:
```
    hdwSize = 2 -- should be positive
    -- 200 coins split among hdwSize users. Should be small sum enough
    -- to avoid making wallets slot leaders.
    hdwDistr = FlatStakes (fromIntegral hdwSize) (mkCoin 200)
```

So there are two addresses each holding 100 coins. It's expected to be much smaller than the total stake in the system.
It's fine initially, but each transaction will give these HD addresses same stake as other genesis addresses. It's quite bad, but not critical.
And it will be solved by weights (HD addresses will have small weights which is already implemented).